src/util/any.hpp

```diff
@@ -0,0 +1,269 @@
+    }
+
+    void swap(any& other) noexcept {
+        std::swap(other.state_, state_);
+    }
+
+    bool has_value() const noexcept {
+        return (bool)state_;
+    }
+
+    const std::type_info& type() const noexcept {
+        return state_->type();
+    }
+
+private:
+
+    struct interface {
+        virtual ~interface() = default;
+        virtual const std::type_info& type() = 0;
+        virtual interface* copy() = 0;
+        virtual void* pointer() = 0;
+        virtual const void* pointer() const = 0;
+    };
+
+    template <typename T>
+    struct model: public interface {
+        ~model() = default;
+
+        interface* copy() override {
+            return new model<T>(*this);
+        }
+
+        const std::type_info& type() override {
+            return typeid(T);
+        }
+
+        model(const T& other):
+            value(other)
+        {}
+
+        model(T&& other):
+            value(std::move(other))
+        {}
+
+        void* pointer() {
+            return &value;
+        }
+
+        const void* pointer() const {
+            return &value;
+        }
+
+        T value;
+    };
+
+    std::unique_ptr<interface> state_;
+
+protected:
+
+    template <typename T>
+    friend const T* any_cast(const any* operand);
+
+    template <typename T>
+    friend T* any_cast(any* operand);
+
+    template <typename T>
+    T* unsafe_cast() {
+        return reinterpret_cast<T*>(state_->pointer());
+    }
+
+    template <typename T>
+    const T* unsafe_cast() const {
+        return reinterpret_cast<T*>(state_->pointer());
+    }
+};
+
+namespace impl {
+    template <typename T>
+    using any_cast_remove_qual = typename
+        std::remove_cv<typename std::remove_reference<T>::type>::type;
+
+    template <typename ValueType, typename U=any_cast_remove_qual<ValueType>>
+    using test_any_cast_constructable = typename
+        std::conditional<
+               std::is_constructible<ValueType, const U&>::value
+            && std::is_constructible<ValueType, U&>::value
+            && std::is_constructible<ValueType, U>::value,
+            std::true_type,
+            std::false_type>::type;
+
+    template<typename T>
+    inline T move_else_copy(any_cast_remove_qual<T>* p, std::true_type) {
+        return std::move(*p);
+    }
+
+    template<typename T>
+    inline T move_else_copy(any_cast_remove_qual<T>* p, std::false_type) {
+        return *p;
+    }
+} // namespace impl
+
+// If operand is not a null pointer, and the typeid of the requested T matches
+// that of the contents of operand, a pointer to the value contained by operand,
+// otherwise a null pointer.
+template<class T>
+const T* any_cast(const any* operand) {
+    static_assert(
+        impl::test_any_cast_constructable<T>::value,
+        "any_cast requires ");
+
+    if (operand==nullptr || !operand->has_value()) {
+        return nullptr;
+    }
+
+    if (operand->type() != typeid(T)) {
+        throw bad_any_cast();
+    }
```

Comment:
The `any_cast<T>(const any*)` overload shouldn't throw (in fact, should be no-except). When `typeid(T)` is not the same as `operand->type()`, it should return `nullptr`.

As an implementation note, you could potentially replace the private `unsafe_cast<T>` method with a 'safe' version that did the `typeid(T)` test and returned `nullptr` on failure.

This applies of course to the non-const version below.